doc/whats_new.rst

```diff
@@ -32,6 +32,7 @@ Changelog
 - :func:`mne_bids.get_matched_empty_room` now implements an algorithm for discovering empty-room recordings that do not have the recording date set as their session, by `Richard Höchenberger`_ (`#421 <https://github.com/mne-tools/mne-bids/pull/421>`_)
 - :func:`write_raw_bids` now adds citations to the README, by `Alex Rockhill`_ (`#463 <https://github.com/mne-tools/mne-bids/pull/463>`_)
 - :func:`make_dataset_description` now has an additional parameter ``dataset_type`` to set the recommended field ``DatasetType`` (defaults to ``"raw"``), by `Stefan Appelhoff`_ (`#472 <https://github.com/mne-tools/mne-bids/pull/472>`_)
+- :func:`mne_bids.copyfiles.copyfile_brainvision` now has ``anonymize`` and ``date`` parameters to control anonymization, by `Stefan Appelhoff`_ (`#475 <https://github.com/mne-tools/mne-bids/pull/475>`_)
```

Comment:
why do you offer two ways of anonymizing? Zen of Python says:

> There should be one-- and preferably only one --obvious way to do it.

we cannot be dogmatic of course, but I fear it will confuse the user. Are there clear differences in the use-cases to do `anonymize=True` in `write_raw_bids` vs in `copyfile_brainvision`? Why does only `copyfile_brainvision` have the `anonymize` parameter? The whole point of API design is to make consistent interfaces that are easy for the user to discover

Reply:
A fair point ... wall of text incoming :-)


> why do you offer two ways of anonymizing?

my main concern with this PR was *not* to offer another way of anonymizing.

The goal was to *stop an unnecessary read-write cycle for BrainVision*. It just doesn't make sense to read BV using MNE-Python, and then rewrite it using `pybv` ... no matter how accurate `pybv` is, it's far simpler to just copy the files and adjust the two lines that need to be anonymized.

(people don't like it when their data gets changed)

Until this PR it was even a bit funny, because users with BV files would use MNE-BIDS and then get a warning "your files are getting converted to BV!!" (which is confusing)


The way I wanted to achieve the above mentioned goal was tp:

1. make minimal adjustments to our "main code" and branching logic (see Alex's comments that I resolved)
1, outsource the main functionality to a short private function

... it just turned out that adding an `anonymize` parameter to `copyfile_brainvision` was the best/cleanest way to achieve this.

---

Having that said --> I *generally* agree with what you say (python zen + api design) *however* in this case I don't think that it hurts to "power up" the copyfile functions.

I know that at least I myself will use this from time to time, when handling BrainVision files. I also took extra care to make the usability clear in the example.

---

Finally: If we were to fully embrace your point, the reasonable thing to do would be to make all copyfile functions private, because we are writing BIDS conversion software here, and not file format utility functions.

Reply:
> Finally: If we were to fully embrace your point, the reasonable thing to do would be to make all copyfile functions private, because we are writing BIDS conversion software here, and not file format utility functions.

But that would be a separate discussion, no?

My concern is that if you modify one `copyfile_xxx` function but not the rest, it will confuse the user _more_ than the confusion caused by a warning which they probably would have ignored. Secondly, I do not want to open a can of worms where folks want to start supporting anonymize for different file formats for the same reasons you mentioned above.

Reply:
> But that would be a separate discussion, no?

yes

> My concern is that if you modify one copyfile_xxx function but not the rest, it will confuse the user more than the confusion caused by a warning which they probably would have ignored.

I already hinted at it in #469 that it'd be simple to do for EDF as well. And I wouldn't be opposed to adding `anonymize` to each copyfile function. But of course, this needs to be done incrementally and a little bit on a need basis.

I *need* MNE-BIDS to *not* read+rewrite my data to BrainVision, because it's already in BrainVision, and I don't want a conversion software to touch my data *content* ... the `anonymize` parameter is the simple solution

> Secondly, I do not want to open a ~can of worms~ **box of joy** where folks want to start supporting anonymize for different file formats for the same reasons you mentioned above.

Why not? if the coding-action is taking place in the `copyfile` functions themselves, there is nothing (almost nothing) we need to change in `write.py`.

The big benefit is that we don't force people to convert their file formats beyond what BIDS is already requiring.

Reply:
> ~can of worms~ **box of joy**

😁



> The big benefit is that we don't force people to convert their file formats beyond what BIDS is already requiring.

This is a very convincing argument, IMHO! +1 from my side…

Reply:
at the outset, I will apologize for being slow and persistent ;-) you just need to explain things to the death for me :-)

> I need MNE-BIDS to not read+rewrite my data to BrainVision, because it's already in BrainVision, and I don't want a conversion software to touch my data content

again, then why do we do this differently for MEG? Are you saying that the `MNE` implementation of writing `fif` files is reliable but the `pybv` implementation of writing brainvision files is not? 

When we anonymize `fif` files in MNE, we read in the whole thing, and write back the whole thing.

Reply:
> at the outset, I will apologize for being slow and persistent ;-) you just need to explain things to the death for me :-)

apology accepted :rofl: 

> again, then why do we do this differently for MEG?

We do it because it's easy for us.

I can imagine that there are also people out there who are annoyed by that. --> HOWEVER I could imagine that for FIF the annoyance is a bit smaller ... because if these people are working with MNE-BIDS, chances are high they work with MNE-Python, ... and over there everything is FIF centric anyhow.

But my main reply to this point would be: IMO it would also be nicer if we didn't *have* to convert to FIF. It would be nicer to let people keep their file format and simply anonymize it --> that's what they want.

> Are you saying that the MNE implementation of writing fif files is reliable but the pybv implementation of writing brainvision files is not?

in a *way* yes ... The FIF reader in MNE is the daily bread and butter at MNE-Python ... it is necessarily the most robust thing in the whole software, because everything depends on it.

`pybv` on the other hand is a niche software that has fewer users, fewer devs, and has been out there for a much shorter time.

I have high hopes that `pybv` will become very robust and nice, but it's not there yet and I don't have the time and energy to single-handedly push it to an equal level of robustness as FIF in MNE-Python.

And with "it's not there yet", I mean the following:

- BrainVision files have a lot of stuff in their headers (see e.g., the impedance PR going on https://github.com/mne-tools/mne-python/pull/7974) --> `pybv` currently more or less discards the whole header beyond the "essential" information.

So yes, `pybv` is good enough for getting the job done when focusing on data accuracy, but it's not equivalent to keeping the original data because of meta-data loss.

> When we anonymize fif files in MNE, we read in the whole thing, and write back the whole thing.

yes, I addressed this with my first points

Reply:
okay I am going to merge this so we don't keep going around in circles. Seems you're more comfortable anonymizing your data this way, but I still think it's better to fix the `pybv` issues than to write workarounds for it. Responses to a couple of points:

> IMO it would also be nicer if we didn't have to convert to FIF. It would be nicer to let people keep their file format and simply anonymize it

yes, but I am specifically talking about the case that does not involve "conversion". You read a fif file in, then you anonymize it and then you write it back as fif. IMO, I/O round trip should not change the file so it loses metadata. You should have tests around this in `pybv` I suppose?

> but it's not there yet and I don't have the time and energy to single-handedly push it to an equal level of robustness as FIF in MNE-Python.

I hear you on this. But now with the workaround, you lose some users + potential contributors.